### ISSUE TITLE
Target multiple vcluster namespaces

### ISSUE THREAD

**Issue by rafaeljusto** [loft-sh/vcluster-sdk#68]
Hi 👋 

I'm working on a scenario where a secret in the host namespace needs to be synchronised to multiple namespaces in the vcluster. To achieve this I'm using the [vCluster Plugin](https://www.vcluster.com/docs/plugins/overview) approach following a similar solution from [this example](https://github.com/loft-sh/vcluster-sdk/blob/02a4201c2ae838eba86e84055f39c3fe0b03b2a9/examples/pull-secret-sync/syncers/pull-secret-sync.go).

My problem is with the [`PhysicalToVirtual`](https://github.com/loft-sh/vcluster-sdk/blob/02a4201c2ae838eba86e84055f39c3fe0b03b2a9/syncer/translator/translator.go#L27-L28) method that only allows returning 1 namespace to be used [for the reconcile request](https://github.com/loft-sh/vcluster-sdk/blob/cd7dfe68c879b9d8a9111d5b2c3d94f4690bfe8f/syncer/syncer.go#L153-L156). Is there an alternative to achieve what I need?

Otherwise, it would be great to have a flexible solution in the SDK, for example:
```go
func (r *syncerController) enqueuePhysical(obj client.Object, q workqueue.RateLimitingInterface) {
	if obj == nil {
		return
	}

	managed, err := r.syncer.IsManaged(obj)
	if err != nil {
		r.log.Errorf("error checking object %v if managed: %v", obj, err)
		return
	} else if !managed {
		return
	}

	// support to target multiple virtual namespaces
	multiple, ok := r.syncer.(interface{
		PhysicalToVirtuals(pObj client.Object) []types.NamespacedName
	})
	if ok {
		for _, name := range multiple.PhysicalToVirtuals(obj) {
			if name.Name != "" {
				q.Add(reconcile.Request{NamespacedName: name})
			}
		}
	}

	name := r.syncer.PhysicalToVirtual(obj)
	if name.Name != "" {
		q.Add(reconcile.Request{NamespacedName: name})
	}
}
```

That way we would be backwards compatible and also allow multiple target namespaces.

**Comment by matskiv** [comment]
Could you describe your scenario in more detail? Maybe there is an easier solution, e.g. [Secret sync in vCluster.Pro](https://www.vcluster.com/pro/docs/secrets/secret-sync).

**Comment by rafaeljusto** [comment]
Sure! Sorry, I should have added more details.

I'm spinning up many applications inside the vCluster, and each application group is kept on a different namespace. Those applications require access to a private Docker registry to retrieve the images. I need to inject the Docker registry secret from the host cluster into each namespace and associate it with the default service account so the apps can spin up.

My company is currently using the open-source version of vCluster, so I cannot rely on pro features.

Thanks in advance!

**Comment by matskiv** [comment]
You could try setting the `serviceAccount.imagePullSecrets` helm value, with the same format as SA imagePullSecrets field. It can reference a Secret in the host namespace where vcluster is installed. And when Pod or ServiceAccount in vcluster doesn't set the pull secret then in theory it should just fall back to this one. :) 

P.S. It won't work if you are enabling this helm value: `sync.serviceaccounts.enabled: true`.

**Comment by matskiv** [comment]
btw, @rafaeljusto let us know whether the `serviceAccount.imagePullSecrets` helm value worked for your use case, so we can close this issue :)

**Comment by matskiv** [comment]
oh, yeah the multi-namespace mode is quite different, that won't work :/ 
Just curious, which CRDs are you syncing that you need multi-ns mode?

**Comment by rafaeljusto** [comment]
> Just curious, which CRDs are you syncing that you need multi-ns mode?

The objective is to replicate the cluster from the production environment in the vCluster. So it's not a big deal to work with a single namespace, but ideally would be nice to have each group of apps on different namespaces, same as in production.

I don't actually need to sync CRDs into the vCluster (at least I didn't get an error yet 😄). That's probably because I'm using the "app of apps" approach in the host cluster to install the apps in the vCluster. The problem is more related to setting up the default service account on each vCluster's namespace associated with a Docker Hub token/credentials secret to pull images from the private registry.

I thought of using the vCluster plugin approach to achieve that. But then got blocked on this limitation of 1:1 mapping. From what I understood, applying a small patch to allow 1:N mapping should allow that, but I could be missing some synchronisation impact on doing that.

**Comment by matskiv** [comment]
Oh, I misunderstood you. It will work with multiple namespaces inside of the vcluster, so you should be good.
I thought that you are talking about this advanced feature - [multi-namespace mode](https://www.vcluster.com/docs/syncer/other_resources/multi_namespace_mode), which you don't need here :) 

About your suggestion about the SDK plugin, we want to cover 80-90% of simple use cases with the SDK, and others are still possible, just not with the structure that SDK provides. And 1: N mapping sounds like a more niche use case, so this would not be included in the SDK unless we hear about more use cases from multiple people.

**Comment by rafaeljusto** [comment]
Hey @matskiv , sorry for the delay. As you mentioned, I tested the 1:N namespace scenario and confirmed that using `serviceAccount.imagePullSecrets`  in the vCluster configuration works!

```yaml
---
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha1
kind: VCluster
metadata:
  name: example
spec:
  controlPlaneEndpoint:
    host: ""
    port: 0
  helmRelease:
    chart:
      name: null
      repo: null
      version: null
    values: |
      serviceAccount:
        create: true
        name: default
        imagePullSecrets:
          - name: dockerconfig
  kubernetesVersion: "v1.27.0"
```

Thank you so much for taking the time to investigate this. ❤️
